one-point-oh.md

```diff
@@ -0,0 +1,146 @@
+
+- [Multiple profiles in Sketches](https://github.com/KittyCAD/modeling-app/issues/1876)
+- Assemblies?
+
+
+### Pre-paying technical debt
+
+**P1**:
+
+- Don't make things worse by adding hacks, especially around the API of KCL
+
+**P3**:
+
+- Supplying user-side IDs to engine
+- code mod API
+
+
+## Language priorities
+
+**P1** (priority ordered):
+
+- reserve keywords (s)
+  - https://github.com/KittyCAD/modeling-app/issues/4486
+  - and give meaning to underscore in identifiers
+- function args
+  - named and optional args (m; *$0*)
+  - only one non-named arg (m; *$1*)
+    - ideally we make it the first arg, but doesn't have to be
+    - could require it be called `self`, but I don't think we need to and that makes back-compat harder
+- simple tidying changes
+  - automatic conversion of round floats to ints, remove the `int()` function (s)
+  - record init: `:` -> `=`  (s)
+  - function decl: remove `=` and `->`, `:` for type annotations (s)
+    - return type syntax?
+  - convert special strings (e.g., `END`, `XY`) to constants (s)
+- std lib
+  - organise into modules (for docs, at least) (l)
+    - potential issues: namespacing, collisions, receivers, name shadowing for std
+  - ensure consistency, remove unnecessary or uncertain functions, check for anything we're badly missing (m)
+    - *depends somewhat on $3, we should do this in any case, but how we do it will depend on whether $3 is achieved*
+- support optionally-implicit `%` in pipelines (m; *$2; depends on $1*)
+
+
+**P2** (priority ordered):
+
+- std lib
+  - remove use of anon records (l; *depends on $0*)
+- immutable model
+  - well-defined rules for when engine calls cause a rendered object, add `construction` flag where necessary (m-l)
+- tagging
+  - support `as` in pipelines (m)
+  - remove `$` syntax (m)
+  - where std lib functions take tags, ensure they're unambiguous and work with `as` variables (m)
+
+
+**P3** (priority ordered):
+
+- std lib
+  - reduce number of functions by using optional args (m-l; *$3; depends on $0*)
+- `var`/`roughly` syntax for unconstrained points/numbers (m-l)
+  - changes how we do 'fixed point' constraints
+
+
+## Docs and education material
+
+Planned for feature freeze period (**P3** until then). We should do some planning before-hand (**P0**).
+
+- std docs
+- guide
+- reference
+- videos and other tutorial stuff
+
+
+## Open questions
+
+- Tags (these might be things we want to change for 1.0), in particular the current approach of collecting all tags into a `tags` field and allowing users of the geometry to refer to internal tags seems sub-optimal:
+  - source names are reified at runtime
+  - construction (including tag names) of an object should be private by default
+  - having a `tags` field violates the principle of tags being just variables
+  - a more principled way to do it is `export` of variables, but this 'works' for block syntax sketches, not so much for pipelines, and not for deeply nested tags
```

Comment:
> a more principled way to do it is `export` of variables

It's true that `$foo` binds `foo` locally, like an output parameter. But the `.tags` field is a whole separate use for tags.

A common use case is having a function that builds a sketch and returns it. If it were a matter of binding variables in a weird way that were accessible in the calling scope, you couldn't call the function multiple times.

Here's an example of this pattern that works today that's quite useful:

```kcl
fn buildShape = (plane, offset) => {
  return startSketchOn(plane)
    |> ...
    |> line([1, 2], %, $important)
    |> ...
    |> close(%)
}

one = buildShape('XZ', 0)
two = buildShape('XY', 22)

extrude(10, one)
extrude(10, two)

// Sketch on each face.
buildMore(one.tags.important)
buildMore(two.tags.important)
```

When I first joined, I thought we wanted to hide the information. That may be true for assemblies. But for a single part, the user needs to be able to access everything in the scene that they could possibly click on.

As it is today, I don't see how it could work with static typing. The tags available in `.tags` would need to be part of the return type of `buildShape()`.

On the other hand, if it were reified at runtime, it could look like this and be a runtime error if it's not defined:

```kcl
buildMore(one.tags['important'])
buildMore(two.tags['important'])
```

A complication, which I think you're calling "dynamic tagging", is that even though Kurt et al have pushed for named tags for stability, others have brought up that if we have loops, then it doesn't really make sense to name a segment inside a loop. It needs an index corresponding to the loop index. So how does that work?

We used to declare the tags with strings, so people would actually do string concatenation between a name and a numeric index to declare them. This is all very dynamic.

Unless you have an idea for a completely different way of doing this, I think it needs to be dynamic.

Reply:
So, I actually think there is a lot that is weird about this way of doing things, the dynamic-ness is the least bad bit. The fact that the name `important` is significant at runtime is weird because it makes variables significant which limits optimisation (or at least forces optimisation to maintain this info), it breaks encapsulation (because if the author changes the name of `important`, it break users of the sketch, which can't be caught statically in all cases because of indexing fields by non-literal strings), furthermore, if the structure of the final solid changes, then the meaning of `important` can change. The fact that `important` tags a line but is used to refer to a surface means that tags are ambiguous and I think this will be hard for users to understand in some cases.

I think there isn't a single solution, but multiple: naming and explicitly making some tags public is part of the solution, and I think these should become fields on the created sketch (so the equivalent of simply `one.important`, though not the extruded solid, IMO). It should also be possible to navigate to each component (or at least collection of components), so `one.extrudedFaces` would give you the set of faces extruded from the sketch. Finally, I think we should have a query system for finding geometry which is the general solution for finding any component which can be clicked, in this case, something like `select(partOf = one, extrudedFrom = one.extrusionSketch.important)`. Obviously, this isn't a 1.0 thing though